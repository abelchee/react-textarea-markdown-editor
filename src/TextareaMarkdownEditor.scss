.tme-container {
  box-sizing: border-box;
  * {
    box-sizing: inherit;
  }

  a {
    text-decoration: none;
  }

  .tme-viewer,
  .tme-textarea {
    border: 1px solid #d4d4d5;
    border-radius: 2px 2px;
    border-top: none;
    padding: 5px;
  }
  .tme-textarea {
    overflow-y: auto;
    width: 100%;
    resize: vertical;
    border-top: none;
    border-top-left-radius: 0;
    border-top-right-radius: 0;
    height: auto;
  }
  .tme-viewer {
    min-height: 20px;
  }

  .tme-menu {
    background-color: #f5f5f5;
    padding: 5px 0;
    border: 1px solid #d4d4d5;
    border-radius: 2px 2px 0 0;
    width: 100%;

    &::after {
      display: table;
      content: ' ';
      clear: both;
    }

    ul.tme-menu-group {
      border-radius: 2px;
      background-color: #fff;
      list-style-type: none;
      margin-block-start: 0;
      margin-block-end: 0;
      padding-inline-start: 0;

      & > li.tme-menu-item {
        &:not(:first-child) {
          border-left-width: 0;
        }
        &:first-child {
          margin-left: 0 !important;
          border-radius: 2px 0 0 2px;
        }
        &:last-child {
          border-radius: 0 2px 2px 0;
        }
        border: 1px solid #ccc;
        user-select: none;
        padding: 5px;
        display: inline-block;
        min-width: 33px;
        text-align: center;

        &:hover {
          border: 1px solid #959595 !important;
          background-color: #f9f9f9;
          cursor: pointer;
        }

        &:active {
          background-image: none;
          background-color: #f2f2f2;
          box-shadow: inset 0 3px 6px rgba(0, 0, 0, 0.1);
          text-shadow: none;
          text-decoration: none;
        }

        &.tme-material-icon {
          padding: 4px 0;
          vertical-align: top;
          img {
            height: 19px;
            width: 23px;
            vertical-align: middle;
          }
        }
      }

      &.left {
        float: left;
        margin-left: 5px;
      }

      &.right {
        margin-right: 5px;
        float: right;
      }
    }
  }
}
