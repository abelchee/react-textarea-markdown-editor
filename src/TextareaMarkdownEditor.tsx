import classNames from 'classnames';
import * as React from 'react';
import { RefObject } from 'react';
import EnhancedTextarea from 'react-enhanced-textarea';
import EditContext from './EditorContext';
import EditorMenu from './EditorMenu';
import { IMarkerGroup } from './type';

export interface ITextareaMarkdownEditor {
  id?: string;
  textareaId?: string;
  className?: string;
  viewerClassName?: string;
  viewerStyle?: object;
  placeholder?: string;
  style?: object;
  textareaStyle?: object;
  rows?: number;
  defaultValue?: string;
  value?: string;
  autoFocus?: boolean;
  readOnly?: boolean;
  onChange?: (value: string) => {};
  onKeyDown?: (event: React.KeyboardEvent) => {};
  onKeyPress?: (event: React.KeyboardEvent) => {};
  doParse: (text: string) => string;
  language?: string;
  markers?: IMarkerGroup[];
  onCopy?: (event: React.ClipboardEvent) => void;
  onCopyCapture?: (event: React.ClipboardEvent) => void;
  onPaste?: (event: React.ClipboardEvent) => void;
  onPasteCapture?: (event: React.ClipboardEvent) => void;
}

interface ITextareaMarkdownEditorState {
  edit: boolean;
  lineMarkers: string[];
  value?: string;
}

class TextareaMarkdownEditor extends React.Component<ITextareaMarkdownEditor, ITextareaMarkdownEditorState> {
  public static defaultProps = {
    language: 'en',
    readOnly: false,
    rows: 5,
  };

  private textareaRef: RefObject<EnhancedTextarea>;

  constructor(props: ITextareaMarkdownEditor) {
    super(props);
    this.textareaRef = React.createRef();
    this.state = {
      edit: !props.readOnly,
      lineMarkers: [],
      value: props.defaultValue,
    };
    this.toggleEdit = this.toggleEdit.bind(this);
    this.onChange = this.onChange.bind(this);
    this.focus = this.focus.bind(this);
    this.mark = this.mark.bind(this);
    this.markLine = this.markLine.bind(this);
    this.registerLineMarker = this.registerLineMarker.bind(this);
    this.markTemplate = this.markTemplate.bind(this);
  }

  public focus() {
    this.textareaRef.current!.focus();
  }

  public append(content: string) {
    this.textareaRef.current!.append(content);
  }

  public mark(prefix: string, suffix: string, defaultText: string, multipleLine?: boolean) {
    if (multipleLine) {
      this.textareaRef.current!.toggleMultipleLineMarker({ prefix, suffix, defaultText });
    } else {
      this.textareaRef.current!.toggleMarker({ prefix, suffix, defaultText });
    }
  }
  public markLine(marker: string) {
    this.textareaRef.current!.toggleLineMarker(marker);
  }
  public registerLineMarker(marker: string) {
    const index = this.state.lineMarkers.indexOf(marker);
    if (index < 0) {
      this.setState({ ...this.state, lineMarkers: [...this.state.lineMarkers, marker] });
    }
  }
  public markTemplate(template: string, multipleLine?: boolean) {
    if (multipleLine) {
      this.textareaRef.current!.toggleMultipleLineTemplate(template);
    } else {
      this.textareaRef.current!.toggleTemplate(template);
    }
  }

  public render() {
    const { readOnly = false } = this.props;
    return (
      <div id={this.props.id} className={classNames('tme-container', this.props.className)} style={this.props.style}>
        <EditContext.Provider
          value={{
            focus: this.focus,
            mark: this.mark,
            markLine: this.markLine,
            registerLineMarker: this.registerLineMarker,
            template: this.markTemplate,
          }}
        >
          <EditorMenu
            markers={this.props.markers}
            readOnly={readOnly}
            language={this.props.language!}
            isEditing={this.state.edit}
            toggleEdit={this.toggleEdit}
          />
          {this.state.edit ? (
            <EnhancedTextarea
              id={this.props.textareaId}
              className="tme-textarea"
              ref={this.textareaRef}
              rows={this.props.rows}
              style={this.props.textareaStyle}
              autoFocus={this.props.autoFocus}
              defaultValue={this.props.value ? undefined : this.state.value}
              value={this.props.value}
              onChange={this.onChange}
              onKeyDown={this.props.onKeyDown}
              onKeyPress={this.props.onKeyPress}
              lineMarkers={this.state.lineMarkers}
              onPaste={this.props.onPaste}
              onPasteCapture={this.props.onPasteCapture}
              onCopy={this.props.onCopy}
              onCopyCapture={this.props.onCopyCapture}
              placeholder={this.props.placeholder}
            />
          ) : (
            <div
              className={classNames('tme-viewer', this.props.viewerClassName)}
              style={this.props.viewerStyle}
              dangerouslySetInnerHTML={{
                __html: this.textareaRef.current ? this.props.doParse(this.textareaRef.current.value) : '',
              }}
            />
          )}
        </EditContext.Provider>
      </div>
    );
  }

  private toggleEdit() {
    this.setState({ ...this.state, edit: !this.state.edit });
  }

  private onChange(value: string) {
    this.setState({ ...this.state, value });
    if (this.props.onChange) {
      this.props.onChange(value);
    }
  }
}

export default TextareaMarkdownEditor;
