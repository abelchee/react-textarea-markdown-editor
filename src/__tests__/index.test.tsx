import * as React from 'react';
import * as renderer from 'react-test-renderer';
import TextareaMarkdownEditor from '../TextareaMarkdownEditor';

describe('TextareaMarkdownEditor', () => {
  test('TextareaMarkdownEditor should initialize as it is', () => {
    const component = renderer.create(
      <TextareaMarkdownEditor
        id="container"
        doParse={it => it}
        textareaId="textarea"
        className="mock-container"
        style={{ width: '100%' }}
        textareaStyle={{ width: '90%' }}
      />,
    );
    const testInstance = component.root;
    const { props } = testInstance.findByType(TextareaMarkdownEditor);
    expect(props.id).toBe('container');

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
