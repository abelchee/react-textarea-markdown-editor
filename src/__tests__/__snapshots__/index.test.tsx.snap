// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaMarkdownEditor TextareaMarkdownEditor should initialize as it is 1`] = `
<div
  className="tme-container mock-container"
  id="container"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="tme-menu"
  >
    <ul
      className="tme-menu-group left"
    >
      <li
        className="tme-menu-item tme-dropdown"
        title="<h1></h1>"
      >
        <span
          className="tme-menu-item-inner"
          onClick={[Function]}
        >
          <b>
            H1
          </b>
        </span>
        <span
          className="tme-dropdown-arrow"
          onClick={[Function]}
        >
          <img
            alt=""
            src="test"
          />
        </span>
        <div
          className="tme-dropdown-content"
        >
          <ul>
            <li
              className="tme-menu-item tme-dropdown-item"
              onClick={[Function]}
              title="<h1></h1>"
            >
              <span
                className="tme-menu-item-inner"
                onClick={[Function]}
              >
                <b>
                  H1
                </b>
              </span>
            </li>
            <li
              className="tme-menu-item tme-dropdown-item"
              onClick={[Function]}
              title="<h2></h2>"
            >
              <span
                className="tme-menu-item-inner"
                onClick={[Function]}
              >
                <b>
                  H2
                </b>
              </span>
            </li>
            <li
              className="tme-menu-item tme-dropdown-item"
              onClick={[Function]}
              title="<h3></h3>"
            >
              <span
                className="tme-menu-item-inner"
                onClick={[Function]}
              >
                <b>
                  H3
                </b>
              </span>
            </li>
          </ul>
        </div>
      </li>
    </ul>
    <ul
      className="tme-menu-group left"
    >
      <li
        className="tme-menu-item tme-dropdown"
        title="Bold"
      >
        <span
          className="tme-menu-item-inner"
          onClick={[Function]}
        >
          <b>
            Bold
          </b>
        </span>
        <span
          className="tme-dropdown-arrow"
          onClick={[Function]}
        >
          <img
            alt=""
            src="test"
          />
        </span>
        <div
          className="tme-dropdown-content"
        >
          <ul>
            <li
              className="tme-menu-item tme-dropdown-item"
              onClick={[Function]}
              title="Bold"
            >
              <span
                className="tme-menu-item-inner"
                onClick={[Function]}
              >
                <b>
                  Bold
                </b>
              </span>
            </li>
            <li
              className="tme-menu-item tme-dropdown-item"
              onClick={[Function]}
              title="Italic"
            >
              <span
                className="tme-menu-item-inner"
                onClick={[Function]}
              >
                <i>
                  Italic
                </i>
              </span>
            </li>
            <li
              className="tme-menu-item tme-dropdown-item"
              onClick={[Function]}
              title="Strikethrough"
            >
              <span
                className="tme-menu-item-inner"
                onClick={[Function]}
              >
                <del>
                  Strikethrough
                </del>
              </span>
            </li>
            <li
              className="tme-menu-item tme-dropdown-item"
              onClick={[Function]}
              title="Blockquote"
            >
              <span
                className="tme-menu-item-inner"
                onClick={[Function]}
              >
                Blockquote
              </span>
            </li>
            <li
              className="tme-menu-item tme-dropdown-item"
              onClick={[Function]}
              title="Inline code"
            >
              <span
                className="tme-menu-item-inner"
                onClick={[Function]}
              >
                Inline code
              </span>
            </li>
            <li
              className="tme-menu-item tme-dropdown-item"
              onClick={[Function]}
              title="Code"
            >
              <span
                className="tme-menu-item-inner"
                onClick={[Function]}
              >
                Code
              </span>
            </li>
          </ul>
        </div>
      </li>
    </ul>
    <ul
      className="tme-menu-group left"
    >
      <li
        className="tme-menu-item"
        title="Unordered List"
      >
        <span
          className="tme-menu-item-inner"
          onClick={[Function]}
        >
          <img
            alt=""
            src="test"
          />
        </span>
      </li>
      <li
        className="tme-menu-item"
        title="Ordered List"
      >
        <span
          className="tme-menu-item-inner"
          onClick={[Function]}
        >
          <img
            alt=""
            src="test"
          />
        </span>
      </li>
    </ul>
    <ul
      className="tme-menu-group left"
    >
      <li
        className="tme-menu-item"
        title="Link"
      >
        <span
          className="tme-menu-item-inner"
          onClick={[Function]}
        >
          <img
            alt=""
            src="test"
          />
        </span>
      </li>
    </ul>
    <ul
      className="tme-menu-group right"
    >
      <li
        className="tme-menu-item tme-link"
      >
        <a
          href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet"
          target="_blank"
        >
          <img
            alt=""
            src="test"
          />
        </a>
      </li>
      <li
        className="tme-menu-item"
        onClick={[Function]}
        title="Preview"
      >
        <span
          className="tme-menu-item-inner"
        >
          <img
            alt=""
            src="test"
          />
        </span>
      </li>
    </ul>
  </div>
  <textarea
    autoComplete="off"
    autoFocus={false}
    className="tme-textarea"
    id="textarea"
    onChange={[Function]}
    onKeyDown={[Function]}
    rows={5}
    style={
      Object {
        "width": "90%",
      }
    }
    wrap="virtual"
  />
</div>
`;
